### ISSUE TITLE
Support Unicode Identifiers

### ISSUE THREAD

**Issue by bd82** [jhipster/prettier-java#116]
## The Problem
Java Identifiers may contain unicode letters, for example:
```java
// my name in Hebrew as the Identifier and its spelling in English as the value.
String שחר = "shahar";
```

See the relevant section in the Java Specifications:
http://m/javashttps://docs.oracle.coe/specs/jls/se11/html/jls-3.html#jls-3.8

The specs define the allowed unicode characters **indirectly** by using the results
of the Java methods:
 - Character.isJavaIdentifierStart(int)
-  Character.isJavaIdentifierPart(int)

These methods are naturally not accessible by default in a Java**Script** run time.

### Represent the unicode character rules as a regular JavaScript RegExp
This is what the Acorn & Esprima Java**Script** parsers do which creates some seriously large regExps:
- https://github.com/acornjs/acorn/blob/d38536b08e0b8a84b7e871bfc2c7dbd7a1c24d90/acorn/src/identifier.js#L24-L31
- https://github.com/jquery/esprima/blob/16aa6f168afbedb97703c896588fc05e9e56ce72/src/character.ts#L1-L8

Assuming the Java unicode rules are similar in complexity to the JavaScript unicode rules
We will be looking at creating (generating?) very large regexps...

### Implement the isJavaIdentifierXXX in JavaScript and use Custom Token Patterns.

A better solution would probably be to avoid using regExps.
Instead Chevrotain can define tokens using arbitrary JavaScript code as well.
This feature is called [Custom Token Patterns](http://sap.github.io/chevrotain/docs/features/custom_token_patterns.html)

What this means is that if we can [translate/compile/access] the Character.isJavaIdentifierXXX
in a JavaScipt runtime, we could implement the unicode Identifiers in a much cleaner fashion
without adding the abstraction of regExps which in this case seems to mostly add complexity.

**Comment by bd82** [comment]
@jdubois I think this is a good candidate for the bounty program.
Should I just add the $$$ label to it?

**Comment by Hawkurane** [comment]
Hello @bd82, I've managed to implement unicode identifiers through a very long and ugly regexp.
It is certainly possible to go through a Custom Token Pattern, but it would still end in matching very but shorter and still ugly regexps. What do you think we should do?

**Comment by clement26695** [comment]
Close this as it is resolved in #155, #170 and #171 (correct me if I'm wrong)
